src/lib.rs

```diff
@@ -0,0 +1,308 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+pub mod rpc;
+pub mod utransport;
+
+use protobuf::{Enum, Message};
+use std::io::Write;
+use std::{
+    collections::HashMap,
+    sync::{atomic::AtomicU64, Arc, Mutex},
+};
+use up_rust::{
+    uprotocol::{UAttributes, UCode, UMessage, UPayloadFormat, UPriority, UStatus, UUri},
+    uri::serializer::{MicroUriSerializer, UriSerializer},
+};
+use zenoh::{
+    config::Config,
+    prelude::r#async::*,
+    queryable::{Query, Queryable},
+    sample::{Attachment, AttachmentBuilder},
+    subscriber::Subscriber,
+};
+
+pub type UtransportListener = Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>;
+
+pub struct ZenohListener {}
+pub struct UPClientZenoh {
+    session: Arc<Session>,
+    // Able to unregister Subscriber
+    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    // Able to unregister Queryable
+    queryable_map: Arc<Mutex<HashMap<String, Queryable<'static, ()>>>>,
+    // Save the reqid to be able to send back response
+    query_map: Arc<Mutex<HashMap<String, Query>>>,
+    // Save the callback for RPC response
+    rpc_callback_map: Arc<Mutex<HashMap<String, Arc<UtransportListener>>>>,
+    callback_counter: AtomicU64,
+}
+
+impl UPClientZenoh {
+    /// # Errors
+    /// Will return `Err` if unable to create Zenoh session
+    pub async fn new(config: Config) -> Result<UPClientZenoh, UStatus> {
+        let Ok(session) = zenoh::open(config).res().await else {
+            return Err(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                "Unable to open Zenoh session",
+            ));
+        };
+        Ok(UPClientZenoh {
+            session: Arc::new(session),
+            subscriber_map: Arc::new(Mutex::new(HashMap::new())),
+            queryable_map: Arc::new(Mutex::new(HashMap::new())),
+            query_map: Arc::new(Mutex::new(HashMap::new())),
+            rpc_callback_map: Arc::new(Mutex::new(HashMap::new())),
+            callback_counter: AtomicU64::new(0),
+        })
+    }
+
+    // TODO: Workaround function. Should be added in up-rust
+    fn get_uauth_from_uuri(uri: &UUri) -> Result<String, UStatus> {
```

Comment:
Would be good to review [Daniel's PR](https://github.com/eclipse-uprotocol/up-rust/pull/54) to make sure this fits what you'll need.